### ISSUE TITLE
all: create a console

### ISSUE THREAD

**Issue by robpike** [upspin/upspin#312]
Daemons such as upspinfs and cacheserver are usually started automatically and run in the background, where any errors or diagnostics they print are never seen. There should be an automatic easy-to-discover Upspin console where one can check for such messages and see other status information, such as what files are being pushed from the cache to the cloud.

Maybe make it available through a menubar item.

**Comment by adg** [comment]
Things I'm interested in learning from the console (I think we're naming this Augie?), in my estimated order of difficulty:

- [ ] that cacheserver is running and healthy
- [ ] that upspinfs is running and healthy
- [ ] the configuration of each of those servers
- [ ] a scrollback of logs from each of those servers
- [ ] the length of the cacheserver writeback queue and the size of those objects
- [ ] the rate (kbps) at which cacheserver is putting objects to remote stores

How does that sound as a basic feature set?

**Comment by adg** [comment]
@andybons started this but he is (totally reasonably) somewhat distracted at the moment. If anyone with macOS dev experience wants to get involved, please let us know.

**Comment by FrankReh** [comment]
I'm assuming protobufs would be used for communication.  

What are the thoughts on the use model? One per machine, one per user logged in, or something else even more granular? What sort of startup and detection is warranted? Is it started, like the cacheserver, when a user config calls for it and it is not already running? Would cacheserver and upspinfs start/reconnect with the console each time they started up automatically? Does the console daemon find them or do they find the console daemon?

Is there any history to be stored in memory beyond what is already being written to local disk, or aside from polling the health status, are logs pulled when the user requests them?

Does it start off with a target audience of developers of upspin on macs or all upspin macOS users? Maybe a normal go background daemon is called for, that can run on any upspin supported platform, doing essentially these functions, and a smaller macOS project for the status bar, menus and pop-up window aspects.  The background daemon could have a command line frontend for getting the same information, perhaps even supporting a kind of tail -f option.  This kind of health and status information seems useful to have over an ssh connection or on any desktop running upspin.

**Comment by FrankReh** [comment]
JSON blobs gotten from the expvar package sounds good.

Before "Augie" is responsible for managing various upspin services, would it be useful if it more simply monitored them?  It could be configured from a file and poll given local ports and ranges of local ports.

**Comment by adg** [comment]
> Before "Augie" is responsible for managing various upspin services, would it be useful if it more simply monitored them? 

This is a fine way to get started, I think.

**Comment by FrankReh** [comment]
Are there any open google projects displaying expvar data in the browser I could draw inspiration from? If not, any other open source projects displaying expvar you think it's okay I look at given the licensing concerns?

**Comment by adg** [comment]
@FrankReh not that I can think of, off the top of my head. What kind of example are you looking for?

**Comment by FrankReh** [comment]
I thought the authors of memstats or expvar might have written some javascript already for displaying the GC info in a browser graphically. I didn't want to be reinventing the wheel if a graphical representation of the memstats info had already been established.

**Comment by gbbr** [comment]
I wrote a package that displays various runtime information in the browser quite a while ago. Not my best piece of work but perhaps it helps: https://github.com/gbbr/memstats

**Comment by SCKelemen** [comment]
If you guys still need someone to write and test for MacOS, I can help. Are there any specific, bite-sized chunks I can cut my teeth on? I've written quite a bit of Go, but I'm new to this project.

**Comment by FrankReh** [comment]
I'm making good progress but I've had to stub out the connection to a unix domain socket for now. I'm using curl to poll the unix sockets but hope that long term a cleaner solution presents itself. The whole point about using them is to use fewer resources on the machine, so having to fire up a child process every poll cycle for each daemon is annoying. The swift standard library comes with an HTTP over TCP implementation but not HTTP over unix domain socket - probably in the future it will. I've thought about manually trying to connect to the socket, sending the GET is trivial, but I haven't figured out how to know when the stream is complete. Probably looking through the golang HTTP code would give us the clue we needed.
I'm keeping a history of poll results in the data, up to a point, and can show how parameters are changing from poll to poll, but haven't decided how to use that information for things like the standard deviation to help highlight outliers.
And since they were already there, the cmdline and memstats have been my primary focus. The console was to show a lot more information than that, log details for one, and cache write-back status for another.

**Comment by adg** [comment]
@FrankReh how about adding a normal TCP HTTP listener to the cacheserver for the exported variables?

**Comment by SCKelemen** [comment]
Can you identify any pieces that I could assist with, given that I am not familiar with this codebase?

**Comment by FrankReh** [comment]
@adg That's what I'm doing - that part works great. For my current tests, I start up upbox, and another upspin client manually having changed the cache server config to use a TCP loopback address - this involved editing the config file that upbox created automatically for the primary user. This is just a temporary problem. Calling curl every few seconds will be okay to get a few folks going I hope. The final solution should be able to use IPC sockets though. Both upbox and upspinfs assume IPC sockets. It could be something I or someone else gets to later. And maybe it will be supported natively in an upcoming Swift version - would be nice for the macOS and Linux desktops which Swift seems to be targeting. So far Go and Node.js do. (And there is no issue with localhost and DNS resolution.)

Apple's documentation on URLSession, which is what I'm using for the TCP HTTP connections, has the tantalizing statement that 'You can also add support for your own custom networking protocols and URL schemes using URLProtocol'. But opening a filesystem named socket, without using C, and without using third party libraries like IBM's BlueSocket project, is for a later time.

**Comment by adg** [comment]
@SCKelemen I don't think this is the most accessible issue for a newbie right now. @FrankReh is already working on some of the UI stuff, and the corresponding cacheserver exported variable work is in progress by @presotto. I'm not sure how we can carve this up further at this point. 🙁

**Comment by FrankReh** [comment]
Okay to target macosx10.12 (Sierra) and newer or will some folks require 10.10 or 10.11 (Yosemite or El Capiton)?

**Comment by robpike** [comment]
I think it's fine to target whatever works best for you to get started.

**Comment by adg** [comment]
I've been thinking about this more.

Ideally we want to have Upspin consoles for each major platform (macOS, Windows, Linux), but we should try to avoid duplication of effort across those consoles. In a perfect world, each of those consoles would just be a thin veneer over a common codebase that manages Upspin processes and collects data from them.

To that end, I propose implementing `upspin-warden`, a daemon that runs on the client machine. Its responsibilities would include:

- starting and re-starting `upspinfs`, `cacheserver`, and any other local Upspin daemons (such as the new accessor that I just added to exp),
- watching the log output of those processes for errors,
- collecting telemetry data for those processes,
- making the collected logs and telemetry data available to the various console apps, through some unified API,
- possibly more, but this seems like a good start.

Any feedback on this idea is welcome. @FrankReh would this make your job easier? I'm assuming it would.

**Comment by FrankReh** [comment]
I like the idea of keeping anything that could compromise the security of the upspin client information on the go side. Keeping the upspin client package all go and all go gettable and all debuggable the normal go way is just a win all the way around. And it means all three platforms are supported. And it means the client package (upspinfs, cacheserver and others) can run headless. Keeping the graphical UI as an optional component will make everyone's life easier. In addition, I like being able to restart the graphical UI without having the client package restart, so the telemetry information, which sometimes can take minutes or hours to accumulate, is already there when the UI is recompiled and restarted.

This doesn't preclude the windowing UI having the ability to start the client package as a single child process, like upbox for example, and then detaching from it. That would allow the windowing UI to know how the upspin clients had been started, so less guessing or configuration involved, if the UI wants to restart and reenter the fray.

**Comment by FrankReh** [comment]
I have some things working now in a windowing program to display the results of polling TCP and IPC endpoints for /debug/vars information. But the display is heavy on the memstats side and light on the rest - because there isn't much to the rest. The only health metric it has so far is the successful poll and parsing as JSON.

What else might contribute to the health status? And maybe the upspin warden would be responsible for deterring the health of things.

The malloc and GC information is probably not very interesting to upspin developers and users. The cmdline args and throughput numbers of cacheserver don't seem like enough to warrant releasing the first cut of the program.

**Comment by adg** [comment]
> This doesn't preclude the windowing UI having the ability to start the client package as a single child process, like upbox for example, and then detaching from it. 

I'd anticipated that the various GUI consoles would start upspin-warden and keep it running. Then the warden would be responsible for running the rest. Is that what you are getting at here? 

> The malloc and GC information is probably not very interesting to upspin developers and users.

It's a starting point, at least. We need to wait on some pending cachserver refactoring before we can put more interesting hooks in.

> And maybe the upspin warden would be responsible for deterring the health of things.

I think so yes.

**Comment by adg** [comment]
I sent out a CL for the start of the upspin-warden daemon: https://upspin-review.googlesource.com/16560

**Comment by frou** [comment]
Perhaps this approach is of interest for a macOS menubar item: https://github.com/matryer/bitbar

**Comment by FrankReh** [comment]
@frou, Thanks but I've got quite a bit working now and will put the first version out soon. After that, I'll be happy to learn more about what others have done in this regard.

**Comment by FrankReh** [comment]
The repo for this macOS app is available. https://github.com/FrankReh/swift-expvar

It polls IPC and TCP socket endpoints for expvar data to display.

It uses the status bar for displaying the overall health of the endpoints, and providing access to a menu. There are windows for looking at the expvar data in various ways.

It builds from the macOS command line using the swift 4 compiler and package manager.

Barring any other immediate feedback, the next step is to incorporate the ability to start a user defined process such as the upspin-warden.

**Comment by FrankReh** [comment]
Should we put our warden questions here or should it have its own issue to track things?

How does the warden start upspinfs? My reading of upspinfs is that its last argument is the mount point but so far warden isn't designed to take such an argument.

Will the config file be able to define the arguments to upspinfs?

**Comment by FrankReh** [comment]
The swift-expvar GUI I hope is useful to demonstrate a little of what is possible but it isn't right as it stands. For an end user on a mac that only wants to see an overall status of the upspin system on their machine and for some cache and fuse details, it is overkill with too much uninteresting detail and too many window options (and doesn't yet show any cache or fuse details). For an upspin developer, it doesn't yet show the logs that warden is keeping.

Maybe two GUIs are called for, or one that defaults to being end-user interesting with an option to show more details.

Warden is now in exp so I want to integrate better with that.

Warden doesn't parse the logs, nor make HTTP queries on the processes it has started. Should warden be responsible for making the judgement as to the status or health of the system so it is done consistently across platforms? Or maybe the goal should be that all status/health can be gleaned from keywords found in the log stream; keywords like warning and error.

**Comment by presotto** [comment]
I added specifying upspinfs' mount point as a config file flag at adg's
request specifically for the warden.

**Comment by FrankReh** [comment]
Thanks. I see it now.

**Comment by FrankReh** [comment]
Sometimes I'm interested in the upspin.Config options set that a process is running with and would like to see it from the GUI.

More daemons could use the expvar package and add a representation of their Config to it. Or daemons could log their Config in a way that warden would collect them as they started.
